[core] Add template parameter for RPC auth in GCS

vitsai opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

This change adds a parameter to the ServerCall object to specify whether to validate the metadata on incoming RPCs. ServerCall objects will by default reject incorrect but not missing metadata. Validation occurs after a ServerCall is popped from its completion queue. By setting the parameter, a GRPC server has the options of STRICT (rejecting incorrect and missing metadata), LAZY (rejecting incorrect but not missing metadata), and NO_AUTH (no authentication).

With this change, GCS service handlers now populate their completion queues with LAZY ServiceCalls, while all others populate with NO_AUTH. 

## Why are these changes needed?
This prevents Ray components from reconnecting to GCS after a restart when GCS is not persisted.

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

iycheng commented on src/ray/gcs/gcs_client/test/gcs_client_test.cc:
```diff
@@ -886,6 +885,7 @@ TEST_P(GcsClientTest, TestGcsTableReload) {
 
   // Restart GCS.
   RestartGcsServer();
+  RAY_CHECK_OK(gcs_client_->Connect(*client_io_service_));
```
Why do we need to reconnect here? This test only works with Redis, so it should just work. The token should be the same as before.

iycheng commented on src/ray/gcs/gcs_client/test/gcs_client_test.cc:
```diff
@@ -1009,13 +1010,15 @@ TEST_P(GcsClientTest, TestGcsAuth) {
   RestartGcsServer();
   auto node_info = Mocker::GenNodeInfo();
 
+  // EXPECT_FALSE(RegisterNode(*node_info));
```
Remove?

iycheng commented on src/ray/gcs/gcs_server/gcs_server.h:
```diff
@@ -14,6 +14,8 @@
 
 #pragma once
 
+#include <atomic>
```
Remove?

iycheng commented on src/ray/rpc/server_call.h:
```diff
@@ -194,21 +204,54 @@ class ServerCallImpl : public ServerCall {
+      }
+    } else {
+      if (!cluster_id_.IsNil()) {
+        RAY_LOG_EVERY_N(WARNING, 100)
+            << "Unexpected cluster ID in server call! " << cluster_id_;
```
Maybe change it to RAY_LOG_EVERY_MS  and give it 5s ? I feel if something goes wrong, this logs might be everywhere.

iycheng commented on src/ray/rpc/server_call.h:
```diff
@@ -194,21 +204,54 @@ class ServerCallImpl : public ServerCall {
-      SendReply(Status::Invalid("HandleServiceClosed"));
+      if (auth_success) {
+        SendReply(Status::Invalid("HandleServiceClosed"));
+      } else {
```
Should we just treat it as ServiceClosed? No strong opinion here.

iycheng commented on src/ray/rpc/server_call.h:
```diff
@@ -223,18 +266,24 @@ class ServerCallImpl : public ServerCall {
-        });
+    if (!auth_success) {
+      boost::asio::post(GetServerCallExecutor(),
+                        [this]() { SendReply(Status::AuthError("WrongClusterToken")); });
```
Maybe add more details about what could be the cause. Somehting like a raylet accessing the GCS which doesn't belong to that cluster?
It should be useful for the users to debug.

iycheng reviewed (commented):
The PR overall looks good to me. We probably should add some end2end test when enable it.

iycheng commented on src/ray/gcs/gcs_server/gcs_server.h:
```diff
@@ -172,6 +174,8 @@ class GcsServer {
   /// Initialize monitor service.
   void InitMonitorServer();
 
+  friend class ray::GcsClientTest;
```
Can we use FRIEND_TEST from gtest?

vitsai replied:
I guess the particular method that uses this, `GcsClientTest::StampContext`, is called in 9 different tests, so having declaration and `FRIEND_TEST` for each individual one might be a bit more code clutter

iycheng reviewed (commented):
Overall I think it looks ok.
Can we add some negative tests for this? We should also add python e2e negative tests to make sure it's working e2e.

vitsai commented:
Added a negative E2E test in Python. The C++ test checks failure before checking success for the non-persisted case.

jjyao commented on python/ray/_private/worker.py:
```diff
@@ -1277,6 +1277,7 @@ def init(
             and the API is subject to change.
         _node_name: User-provided node name or identifier. Defaults to
             the node IP address.
+        _enable_cluster_auth: Whether to enable cluster authentication.
```
Can we just use a normal ray config var (ray_config_def.h)?

vitsai replied:
ah...

vitsai replied:
done

iycheng commented on python/ray/tests/test_gcs_fault_tolerance.py:
```diff
@@ -828,6 +829,37 @@ def f():
     wait_for_pid_to_exit(gcs_server_pid, 10000)
 
 
+@pytest.mark.skipif(
+    enable_external_redis(), reason="Only valid when started without external redis"
+)
```
Should we remove this and if external redis is enabled, we just make sure the raylet is healthy?

vitsai replied:
done

iycheng commented on python/ray/tests/test_gcs_fault_tolerance.py:
```diff
@@ -828,6 +829,37 @@ def f():
+    enable_external_redis(), reason="Only valid when started without external redis"
+)
+@pytest.mark.parametrize(
+    "ray_start_with_env_vars",
```
you don't need a separate fixture for this, search `generate_system_config_map` in this file and follow that pattern.

vitsai replied:
done

jjyao commented on src/ray/raylet/node_manager.cc:
```diff
@@ -551,14 +551,12 @@ ray::Status NodeManager::RegisterGcs() {
-                      << "GCS is not backed by a DB and restarted or there is data loss "
-                      << "in the DB.";
-                }
+              if ((status.ok() && !alive) || status.IsAuthError()) {
```
I think AuthError needs a different fatal message, it's different from the existing case where the node belongs to this cluster but it's dead.

I guess when AuthError happens, there must be a bug right

vitsai replied:
done

jjyao commented on src/ray/rpc/gcs_server/gcs_rpc_client.h:
```diff
@@ -87,11 +87,12 @@ class Executor {
 /// TODO(sang): Fix it.
-#define VOID_GCS_RPC_CLIENT_METHOD(                                                      \
-    SERVICE, METHOD, grpc_client, method_timeout_ms, SPECS)                              \
+#define _VOID_GCS_RPC_CLIENT_METHOD(                                                     \
+    SERVICE, METHOD, grpc_client, method_timeout_ms, SPECS, IS_INSECURE)                 \
```
Actually why do we need to change client side?

vitsai replied:
removed, can use Nil

jjyao commented on src/ray/rpc/gcs_server/gcs_rpc_server.h:
```diff
@@ -141,7 +141,7 @@ namespace rpc {
 
 // TODO(vitsai): Set auth for everything except GCS.
```
Is this TODO still needed?

vitsai replied:
done

jjyao commented on src/ray/rpc/gcs_server/gcs_rpc_server.h:
```diff
@@ -141,7 +141,7 @@ namespace rpc {
 // TODO(vitsai): Set auth for everything except GCS.
 #define INTERNAL_KV_SERVICE_RPC_HANDLER(HANDLER) \
-  RPC_SERVICE_HANDLER(InternalKVGcsService, HANDLER, -1)
+  RPC_SERVICE_HANDLER_CUSTOM_AUTH(InternalKVGcsService, HANDLER, -1, AuthType::NO_AUTH)
```
Internal KV should also have auth?

vitsai replied:
done, adding it back if there is an unexpected test failure

jjyao commented on src/ray/rpc/server_call.h:
```diff
@@ -29,6 +29,13 @@
+enum class AuthType {
+  NO_AUTH,      // Do not authenticate (accept all).
+  LAZY_AUTH,    // Accept missing token, but reject wrong token.
+  STRICT_AUTH,  // Reject missing token and wrong token.
```
I think this is dead code at least for now.

vitsai replied:
removed, we can add it back when we use it

vitsai commented:
Major changes in response to comments:
- Removed client-side template since it is redundant with passing Nil.
- Removed STRICT auth since it is currently unused.

vitsai commented on src/ray/gcs/gcs_server/gcs_server.cc:
```diff
@@ -15,6 +15,7 @@
 #include "ray/gcs/gcs_server/gcs_server.h"
 
 #include <fstream>
+#include <future>
```
done

zhe-thoughts reviewed (approved):
I think this PR qualifies as a release blocker. 

But @jjyao or @iycheng , as real Ray Committers, could you give me a green light when all necessary tests have passed? Then I will merge

jjyao commented on src/ray/rpc/server_call.h:
```diff
@@ -223,18 +257,27 @@ class ServerCallImpl : public ServerCall {
+    if (!auth_success) {
+      boost::asio::post(GetServerCallExecutor(), [this]() {
+        SendReply(
+            Status::AuthError("WrongClusterToken: Perhaps the client is accessing GCS "
```
```suggestion
            Status::AuthError("Wrong cluster id: Perhaps the client is accessing GCS "
```

jjyao commented on src/ray/rpc/gcs_server/gcs_rpc_server.h:
```diff
@@ -382,7 +381,11 @@ class NodeInfoGrpcService : public GrpcService {
+        NodeInfoGcsService,
+        GetClusterId,
+        RayConfig::instance().gcs_max_active_rpcs_per_handler(),
+        AuthType::LAZY_AUTH);
```
This should be NO_AUTH?

vitsai replied:
I changed it, but I'm not sure there's a clear answer to whether we want to allow raylets that originally belonged to a different cluster to rejoin a new cluster.

vitsai replied:
Actually, let's keep it this way to enforce that raylets can only have one cluster ID in their life cycle?

jjyao commented on src/ray/rpc/server_call.h:
```diff
@@ -194,21 +203,46 @@ class ServerCallImpl : public ServerCall {
+        }
+      } else {
+        if (!cluster_id_.IsNil()) {
+          RAY_LOG_EVERY_MS(WARNING, 5000)
```
Should this be RAY_CHECK as well?

vitsai replied:
This one is for the NO_AUTH case

vitsai replied:
(so wrong cluster ID is fine)

vitsai replied:
I guess technically if we don't need it and we have it stored locally, it's not a big deal. I did change the actual warning log to log whether the client has metadata, though.

vitsai replied:
Actually left it as is. There's not much value to checking the client one, and I think periodically logging that the object has a cluster ID that it doesn't need is fine.

jjyao commented on src/ray/rpc/client_call.h:
```diff
@@ -270,6 +274,9 @@ class ClientCallManager {
   /// Get the main service of this rpc.
   instrumented_io_context &GetMainService() { return main_service_; }
 
+  friend class ray::GcsClientTest;
```
What private things we are accessing?

vitsai replied:
StampContext originally accessed cluster_id_ directly, but I've changed ClientCallManager to surface this in a public getter

jjyao commented on src/ray/gcs/gcs_client/test/gcs_client_test.cc:
```diff
@@ -945,9 +960,46 @@ TEST_P(GcsClientTest, TestEvictExpiredDestroyedActors) {
+    return;
+  }
+
+  EXPECT_FALSE(RegisterNode(*node_info));
+  ReconnectClient();
+  EXPECT_TRUE(RegisterNode(*node_info));
```
Could you add some comments for what you are testing?

jjyao reviewed (commented):
lgtm

vitsai commented on src/ray/gcs/gcs_server/gcs_server.h:
```diff
@@ -172,6 +174,8 @@ class GcsServer {
   /// Initialize monitor service.
   void InitMonitorServer();
 
+  friend class ray::GcsClientTest;
```
This one can't, because StampContext still checks client_call_manager_ directly

jjyao replied:
This one `RAY_CHECK(gcs_client_->client_call_manager_)`?

Can we make a public method something like `IsClientCallManagerInitialized()` so we can remove the friend class?

Or is this check needed in the first place?

vitsai commented on python/ray/tests/test_gcs_fault_tolerance.py:
```diff
@@ -828,6 +829,41 @@ def f():
+        # Waiting for raylet to become unhealthy
+        wait_for_condition(lambda: not check_raylet_healthy())
+    else:
+        # Waiting for raylet to become unhealthy
```
fixed comment

vitsai commented on python/ray/tests/test_gcs_fault_tolerance.py:
```diff
@@ -828,6 +829,41 @@ def f():
 
 
+@pytest.mark.parametrize(
+    "ray_start_regular",
```
It will run in the "External redis tests" unless there is a pytest.skipif. An earlier commit had the skipif instead of the alive check for Redis, and an even earlier commit failed the Redis check.

jjyao commented on src/ray/gcs/gcs_client/test/gcs_client_test.cc:
```diff
@@ -115,6 +115,15 @@ class GcsClientTest : public ::testing::TestWithParam<bool> {
   }
 
+  void ReconnectClient() {
+    ClusterID cluster_id = gcs_server_->rpc_server_.GetClusterId();
```
Can we also provide a GetClusterId method so we don't need friend class in gcs_server as well?
